### PR TITLE
extended period syntax

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- redistribute the function `splitString()` from PIConGPU within PMacc
-  add helper function `cointainsStep()`
- add parser function `toTimeSlice()`
- add class TimeSlice
- refactore `PluginConnector`
  - change internal stored type
  - change interface of `setNotificationPeriod()`, change type of period
- use new syntax for checkpoint period definition
- add time slice feature to all PIConGPU plugins
- add documentation

# Tests

- [x] create checkpoints `--checkpoint.period 50,10:12:1`
- [x] phase space plugin with `--e_phaseSpace.period 0:50:10,::`

### CODE REVIEW

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 1]
```diff
@@ -311,7 +310,7 @@ class IsaacPlugin : public ILightweightPlugin
     {
         /* register command line parameters for your plugin */
         desc.add_options()
-            ("isaac.period", po::value< uint32_t > (&render_interval)->default_value(0),
+            ("isaac.period", po::value< uint32_t > (&render_interval),
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: type is now `std::string` for the period but `render_interval` is still set directly here. please check :)

File: include/picongpu/plugins/IsaacPlugin.hpp [hunk 2]
```diff
@@ -366,7 +365,7 @@ class IsaacPlugin : public ILightweightPlugin
         {
             //using an internal variable "render_interval" as notifyPeroid
             //of PIConGPU cannot be changed at runtime
-            notifyPeriod = 1;
+            notifyPeriod = "1";
```
Comment (ax3l) [inline include/picongpu/plugins/IsaacPlugin.hpp]: @theZiz can you confirm this is correct for the ISAAC plugin?
The GUI can change the `render_interval`, right? This is not the same as the `notifyPeriod` I guess and should still work, no? (see: `//using an internal variable "render_interval" as notifyPeroid`)

Please add your thoughts.

File: include/picongpu/plugins/PngPlugin.hpp [hunk 3]
```diff
@@ -102,8 +102,8 @@ namespace picongpu
                 {
                     for (int i = 0; i < (int) slicePoints.size(); ++i) /*!\todo: use vactor with max elements*/
                     {
-                        uint32_t frequ = getValue(notifyPeriod, i);
-                        if (frequ != 0)
+                        std::string frequ = getValue(notifyPeriod, i);
```
Comment (ax3l) [inline include/picongpu/plugins/PngPlugin.hpp]: can you please rename `frequ` to `period` if we are on it? :)

File: include/pmacc/misc/splitString.hpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+     *   source: https://stackoverflow.com/a/28142357
+     *   author: Marcin
+     *   date: Jan 25 '15
+     *
+     * @param input string to split
+     * @param regex separator between two elements
+     */
+    std::vector< std::string > splitString(
+        std::string const & input,
+        std::string const & regex = ","
```
Comment (ax3l) [inline include/pmacc/misc/splitString.hpp]: delimiter?

File: include/pmacc/misc/splitString.hpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+{
+    /** split a string in a vector of strings
+     *
+     * Based on Stack Overflow post:
+     *   source: https://stackoverflow.com/a/28142357
+     *   author: Marcin
+     *   date: Jan 25 '15
+     *
+     * @param input string to split
+     * @param regex separator between two elements
```
Comment (ax3l) [inline include/pmacc/misc/splitString.hpp]: `@param delimiter separator between two elements (supports regex)`

File: include/pmacc/pluginSystem/TimeSlice.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+         * @param str value to set, can be empty
+         */
+        void setValue(uint32_t const idx, std::string const & str)
+        {
+            if(!str.empty())
+            {
+                uint32_t value = std::stoul( str );
+                PMACC_VERIFY_MSG(
+                    !( idx == 2 && value == 0 ),
+                    "Zero is not valid period"
```
Comment (ax3l) [inline include/pmacc/pluginSystem/TimeSlice.hpp]: not **a** valid period

File: include/pmacc/pluginSystem/cointainsStep.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+#include "pmacc/pluginSystem/TimeSlice.hpp"
+
+#include <vector>
+
+
+namespace pmacc
+{
+namespace pluginSystem
+{
+    /** check if a given step is defined within a interval list
```
Comment (ax3l) [inline include/pmacc/pluginSystem/cointainsStep.hpp]: - ~~defined~~
- a**n** interval

File: include/pmacc/pluginSystem/cointainsStep.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @param seqTimeSlices vector with time intervals
+     * @param timeStep simulation time step to check
+     * @return true if step is included in the interval list else false
+     */
+    bool cointainsStep(
```
Comment (ax3l) [inline include/pmacc/pluginSystem/cointainsStep.hpp]: contains ;)

File: include/pmacc/pluginSystem/toTimeSlice.hpp [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+     *
+     * Parse a comma separated list of time slices and creates a vector of TimeSlices.
+     * TimeSlice Syntax:
+     *   - `start:stop:period`
+     *   - a number ``N is equal to `::N`
+     */
+    std::vector< TimeSlice > toTimeSlice( std::string const & str )
+    {
+        std::vector< TimeSlice > result;
+        auto seqOfSlices = misc::splitString(
```
Comment (ax3l) [inline include/pmacc/pluginSystem/toTimeSlice.hpp]: const?

File: include/pmacc/pluginSystem/toTimeSlice.hpp [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            str,
+            ","
+        );
+        for( auto const & slice : seqOfSlices )
+        {
+            auto sliceComponents = misc::splitString(
```
Comment (ax3l) [inline include/pmacc/pluginSystem/toTimeSlice.hpp]: const?

File: include/pmacc/pluginSystem/toTimeSlice.hpp [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                ":"
+            );
+            PMACC_VERIFY_MSG(
+                !sliceComponents.empty( ),
+                std::string( "time slice without a defined element is not allowed" ) + str
+            );
+
+            // id of the component
+            size_t n = 0;
+            bool hasOnlyPeriod = sliceComponents.size() == 1u;
```
Comment (ax3l) [inline include/pmacc/pluginSystem/toTimeSlice.hpp]: const?

File: include/pmacc/simulationControl/SimulationHelper.hpp [hunk 12]
```diff
@@ -134,7 +141,13 @@ class SimulationHelper : public IPlugin
         Environment<DIM>::get().PluginConnector().notifyPlugins(currentStep);
 
         /* trigger checkpoint notification */
-        if (checkpointPeriod && (currentStep % checkpointPeriod == 0))
+        if(
+            !checkpointPeriod.empty() &&
+            pluginSystem::cointainsStep(
```
Comment (ax3l) [inline include/pmacc/simulationControl/SimulationHelper.hpp]: contains

File: include/picongpu/plugins/IsaacPlugin.hpp (hunk 1 above)
Comment (psychocoderHPC) [reply include/picongpu/plugins/IsaacPlugin.hpp]: I will check the behavior in a follow up after this PR

File: include/picongpu/plugins/IsaacPlugin.hpp (hunk 2 above)
Comment (ax3l) [reply include/picongpu/plugins/IsaacPlugin.hpp]: follow-up PR, see comment above

**Comment by psychocoderHPC** [comment]
ready for to merge

**Pull request merged**